packages/react-dom/src/client/ReactDOMFiberComponent.js

```diff
@@ -80,6 +93,235 @@ let normalizeHTML;
 if (__DEV__) {
   getStack = getCurrentFiberStackAddendum;
 
+  const NODE_CONTENT_CLIP_LENGTH = 100;
+
+  const clipStringWithEllipsis = function(str: string, clipAtLength: number) {
+    return (
+      str.substring(0, clipAtLength) + (str.length > clipAtLength ? '…' : '')
+    );
+  };
+
+  getTextContentSignature = function(textContent: string) {
+    // TODO: How do we want to print quotes and other special characters inside string quotes?
```

Comment:
Should we resolve TODOs before getting this in?

Reply:
Yes, I'd love to hear some thoughts from the React team how you'd like the TODOs to be resolved, as solutions vary in volume and complexity.

Reply:
Can you offer some options? I've never thought about this so it would help to get a braindump.

Reply:
RE: displaying special characters in quotes, the braindump:
1. There are two contexts where quoted strings appear: HTML and JS/JSX.
2. In HTML, the special characters are escaped with `&mnemonic;` and `&#hexnumber;` sequences. The implementation of such display may vary from writing a function that prints the char codes, or reusing the browser implementation via set innerHTML/get innerText on an off-screen DOM node.
3. In JavaScript, the special characters are escaped with a backslash and a mnemonic or a number. The implementation may vary from implementing the escaping in code for certain characters to reusing `JSON.stringify`. The latter generates double-quoted strings whereas React's JavaScript style uses single quotes, and there's no standard "JavaScript.stringify" function.
4. Given the above adds some complexity around displaying values on the console, the other solution would be to dump the values as they are.
5. React already prints some values to the console, what is the pattern to do that? Which utility functions are used?

Reply:
What is the input to this function? Is it a string constructed on the client side? A string we read from the DOM? Sometimes one, sometimes the other?

Can you add some tests that specify the current behavior, and then we discuss that?

Reply:
See 👉 https://github.com/facebook/react/pull/12063/commits/e085f69436485b2f134e670da99edf0cbd222e00 for a couple of examples that came to mind.

I could also add a pack of unit tests for these functions specifically, but I wonder if these functions are to be extracted into files, how would you prefer to lay out the files and how to import/require them, and what about having tests that do not use the public API.